internal/terraform/module/module_ops.go

```diff
@@ -440,20 +440,58 @@ func ParseVariables(ctx context.Context, fs ReadOnlyFS, modStore *state.ModuleSt
 
 	// TODO: Avoid parsing if the content matches existing AST
 
-	// TODO: Only parse the file that's being changed/opened, unless this is 1st-time parsing
-
 	// Avoid parsing if it is already in progress or already known
 	if mod.VarsDiagnosticsState[ast.HCLParsingSource] != op.OpStateUnknown && !job.IgnoreState(ctx) {
 		return job.StateNotChangedErr{Dir: document.DirHandleFromPath(modPath)}
 	}
 
-	err = modStore.SetVarsDiagnosticsState(modPath, ast.HCLParsingSource, op.OpStateLoading)
+	var files ast.VarsFiles
+	var diags ast.VarsDiags
+	rpcContext := lsctx.RPCContext(ctx)
+	// Only parse the file that's being changed/opened, unless this is 1st-time parsing
+	if mod.ModuleDiagnosticsState[ast.HCLParsingSource] == op.OpStateLoaded && rpcContext.IsDidChangeRequest() && lsctx.IsLanguageId(ctx, ilsp.Tfvars.String()) {
+		// the file has already been parsed, so only examine this file and not the whole module
+		err = modStore.SetVarsDiagnosticsState(modPath, ast.HCLParsingSource, op.OpStateLoading)
+		if err != nil {
+			return err
+		}
+		filePath, err := uri.PathFromURI(rpcContext.URI)
+		if err != nil {
+			return err
+		}
+		fileName := filepath.Base(filePath)
+
+		f, vdiags, err := parser.ParseVariableFile(fs, filePath)
+		if err != nil {
+			return err
+		}
+
+		existingFiles := mod.ParsedVarsFiles.Copy()
+		existingFiles[ast.VarsFilename(fileName)] = f
+		files = existingFiles
+
+		existingDiags, ok := mod.VarsDiagnostics[ast.HCLParsingSource]
+		if !ok {
+			existingDiags = make(ast.VarsDiags)
+		} else {
+			existingDiags = existingDiags.Copy()
+		}
+		existingDiags[ast.VarsFilename(fileName)] = vdiags
```

Comment:
```suggestion
		f, vDiags, err := parser.ParseVariableFile(fs, filePath)
		if err != nil {
			return err
		}

		existingFiles := mod.ParsedVarsFiles.Copy()
		existingFiles[ast.VarsFilename(fileName)] = f
		files = existingFiles

		existingDiags, ok := mod.VarsDiagnostics[ast.HCLParsingSource]
		if !ok {
			existingDiags = make(ast.VarsDiags)
		} else {
			existingDiags = existingDiags.Copy()
		}
		existingDiags[ast.VarsFilename(fileName)] = vDiags
```
Naming nitpick: Variable names in Go tend to be camelCase.